# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def addTwo(x):
    return x + 2

print(addTwo(4))

# The lambda version...
# Without using a variable to pass in data
print((lambda x: x + 2)(4))

# Using a variable
a_num = 4
print((lambda x: x + 2)(a_num))

6
6
6


#### Saving to a Variable

In [2]:
f_test = lambda x: x + 2

f_test(4)

6

#### Multiple Inputs

In [3]:
#multiple input with no vars
print((lambda x, y, z: x*y*z)(3,5,8))
#multiple inputs with var
x_test = lambda x, y ,z: x*y*z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [4]:
def multiply(f, num):
    '''
     f expects a lambda funciton
     num expects a number
    '''
    return f(num)
multiply(lambda x: x * x, 4)

16

#### Returning a Lambda from a Function

In [5]:
# Regular defined function:
def multiply_tst(num):
    return num * 4

# Function within a function:
def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(returnFunc())
print(f_return(4))

# Lambda function returned from a regular function
def returnLam(b,c):
    return lambda x, a: x + a + b + c
r_lamb = returnLam(4,5)
print(r_lamb(5,5))

<function returnFunc.<locals>.multiply at 0x000001BFF78C8F40>
8
19


#### If Statements within Lambdas

In [6]:
# Lambda x: True if (condition) else False
f_condition = lambda num : num * 2 if num > 10 else num + 2

print(f_condition(8))
print(f_condition(12))
print(f_condition(10))

10
24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [6]:
x = lambda f: f**3
x(6)

216

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [7]:
def squared(a,b):
    if a < 10 and b < 10:
        return a**2, b**2
    else:
        return a, b
numbers = [1,3,5,2,342,235,12,123,122,35,6,2]
num2 = [1,2,3,4,5,6,7,8,8,9,5678]
squaremap = map(squared, numbers, num2)
print(list(squaremap))
 

[(1, 1), (9, 4), (25, 9), (4, 16), (342, 5), (235, 6), (12, 7), (123, 8), (122, 8), (35, 9), (6, 5678)]


#### Using Lambda's with Map

In [9]:
# map(lambda x: do something, iterable)
# Using lambda in map happens inline or on one line
#list(map(lambda x: x * 2, nums))
numbers = [1, 2, 3, 11, 12, 13]  # replace with your list of numbers
more_nums = [4, 5, 6, 14, 15, 16]  # replace with your second list of numbers

print(list(map(lambda x, y: (x ** 2, y ** 2) if x < 10 and y < 10 else (x,y), numbers, more_nums)))

[(1, 16), (4, 25), (9, 36), (11, 14), (12, 15), (13, 16)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [12]:
# Let's assume this is your list
numbers = [1, 2, 3, 4, 5]

# Use map() with a lambda function
result = list(map(lambda x: x*2 - 1, numbers))

print(result)

[1, 3, 5, 7, 9]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [21]:
names = ['Bob', 'Andy', 'Max', 'Evan', 'Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    elif name[0].lower() == 'b':
        return True
    else:
        return False
new_names_list = list(filter(a_names, names))
print(new_names_list)

['Bob', 'Andy', 'Angelica']


#### Using Lambda's with Filter()

In [24]:
new_names_lambda = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lambda)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [8]:

import statistics
listt = [1,2,3,5,1,124,123,23,2,2,3,234,25,62,133,12,23,73,234,12,32,2,2]
m = statistics.mean(listt)
print(m)
f = set(filter(lambda x: True if x >= m else False, listt))
print(f)

49.26086956521739
{133, 73, 234, 123, 124, 62}


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [29]:
from functools import reduce

# reduce(function, iterable)

list_1 = [2, 4, 6, 8, 10]

def addNums(num1, num2):
    return num1 + num2
result_add = reduce(addNums, list_1)

print(result_add)

# Substract a list of numbers
def subtractNums(num1, num2):
    return num1 - num2

result_sub = reduce(subtractNums, list_1)
print(result_sub)

30
-26


#### Using Lambda's with Reduce()

In [30]:
result_lamb = reduce(lambda x, y: x + y, list_1)
print(result_lamb)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [9]:
listt = [1,2,5,2,3,23,6,2,6,2]
a = 1
for x in listt:
    a += a*x
print(a)

4572288


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [31]:
def addNums(num):
    # Here is our base case...
    if num <= 1:
        print("addNums(1) = 1")
        return num
    else:
        print(f"addNums({num}) = {num} + addNums({num -1})")
        return num + addNums(num -1)
addNums(5)

addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1


15

#### Writing a Factorial Function

In [32]:
# 5| = 5 * 4 * 3 * 2 * 1
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1)
factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [10]:

l = [1,2,3,4,5,6,7,8,9,10]
def s(n,l):
    if len(l)==0:
        return 0
    else:
        n = n-l[0]
        l.pop(0)
        s(n,l)
print(s(1,l))

None


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [33]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start # yield keyword denotes a generator
        start += step
for i in my_range(20, start = 2):
    my_generator_value = i
    print(my_generator_value)

my_range(20, start =2)

    

2
4
6
8
10
12
14
16
18


<generator object my_range at 0x00000267170BA400>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [11]:

def r(stop, start, step = -1):
    while stop <=start:
        yield start
        start += step
for i in r(0,start = 10):
    a = i**2
    print(a)
    

100
81
64
49
36
25
16
9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [16]:

# Let's assume this is your list
places = ['Argentina', '', 'San Diego', 'Boston', '', 'New York', '']

# Use filter() to remove all empty strings
filtered_places = list(filter(None, places))

print(filtered_places)


['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [17]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

# Let's assume this is your list
names = ['Andrew P. Garfield', 'David hassELHOFF', 'Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter']

# Use sort() with a lambda function as the key
names.sort(key=lambda name: name.split()[-1].lower())

print(names)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [1]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
places_temps = [('Nashua',32),('Boston',12),('Los Angelos',44),('Miami',29)]

# Use map() with a lambda function to convert temperatures
converted_temps = list(map(lambda x: (x[0], x[1] * 9/5 + 32), places_temps))

print(converted_temps)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [ ]:
def fibonacci(n):
    if n <= 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fibonacci(n-1) + fibonacci(n-2)
    
nt = int(input("How many terms do you want?"))
def a(n):
    if n <= 1:
        return n
    else:
        return a(n-1)+a(n-2)
for i in range(nt):
    print("Iteration",i,": ",a(i+1))
            

# Print the Fibonacci sequence up to the 5th number
for i in range(6):
    print(f"Iteration {i}: {fibonacci(i)}")

How many terms do you want?69
Iteration 0 :  1
Iteration 1 :  1
Iteration 2 :  2
Iteration 3 :  3
Iteration 4 :  5
Iteration 5 :  8
Iteration 6 :  13
Iteration 7 :  21
Iteration 8 :  34
Iteration 9 :  55
Iteration 10 :  89
Iteration 11 :  144
Iteration 12 :  233
Iteration 13 :  377
Iteration 14 :  610
Iteration 15 :  987
Iteration 16 :  1597
Iteration 17 :  2584
Iteration 18 :  4181
Iteration 19 :  6765
Iteration 20 :  10946
Iteration 21 :  17711
Iteration 22 :  28657
Iteration 23 :  46368
Iteration 24 :  75025
Iteration 25 :  121393
Iteration 26 :  196418
Iteration 27 :  317811
Iteration 28 :  514229
Iteration 29 :  832040
Iteration 30 :  1346269
Iteration 31 :  2178309
Iteration 32 :  3524578
Iteration 33 :  5702887
Iteration 34 :  9227465
Iteration 35 :  14930352
Iteration 36 :  24157817
Iteration 37 :  39088169
Iteration 38 :  63245986
Iteration 39 :  102334155
Iteration 40 :  165580141
Iteration 41 :  267914296
